### ISSUE TITLE
script and doc_as_upsert throw exception 

### ISSUE THREAD

**Issue by salimane** [logstash-plugins/logstash-output-elasticsearch#364]
The following is my logstash configuration
```
input {
  rabbitmq {
    host => "localhost"
    port => 5672
    vhost => "/"
    user => "my_user"
    password => "my_pwd"
    queue => "my_queue"
    exchange => "my_exchange"
    key => "my_routing_key"
  }
}
output {
  elasticsearch {
    action => "update"
    doc_as_upsert => true
    script => "ctx._source.count += 1"
    script_lang => "groovy"
    script_type => "inline"
    document_type => "my_type"
    index => "my_index"
    hosts => ["localhost:9200"]
    document_id => "%{id}"
  }
}
```

Basically what i want for every message from rabbitmq : 
- create a new doc in es with the event
- update the doc if it already exists in es
- while updating an existing doc, increase the ``count`` field by 1

Currently with logstash 2.2 and es 1.7.1, the above configuration generates an error : 
```
{:timestamp=>"2016-02-13T20:28:54.526000+0000", :message=>"Attempted to send a bulk request to Elasticsearch configured at '[\"http://localhost:9200/\"]', but an error occurred and it failed! Are you sure you can reach elasticsearch from this machine using the configuration provided?", :client_config=>{:hosts=>["http://localhost:9200/"], :ssl=>nil, :transport_options=>{:socket_timeout=>0, :request_timeout=>0, :proxy=>nil, :ssl=>{}}, :transport_class=>Elasticsearch::Transport::Transport::HTTP::Manticore, :logger=>nil, :tracer=>nil, :reload_connections=>false, :retry_on_failure=>false, :reload_on_failure=>false, :randomize_hosts=>false}, :error_message=>"[400] {\"error\":\"ActionRequestValidationException[Validation Failed: 1: script or doc is missing;2: script or doc is missing;]\",\"status\":400}", :error_class=>"Elasticsearch::Transport::Transport::Errors::BadRequest", :backtrace=>["/opt/logstash/vendor/bundle/jruby/1.9/gems/elasticsearch-transport-1.0.15/lib/elasticsearch/transport/transport/base.rb:146:in `__raise_transport_error'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/elasticsearch-transport-1.0.15/lib/elasticsearch/transport/transport/base.rb:256:in `perform_request'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/elasticsearch-transport-1.0.15/lib/elasticsearch/transport/transport/http/manticore.rb:54:in `perform_request'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/elasticsearch-transport-1.0.15/lib/elasticsearch/transport/client.rb:125:in `perform_request'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/elasticsearch-api-1.0.15/lib/elasticsearch/api/actions/bulk.rb:87:in `bulk'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.4.1-java/lib/logstash/outputs/elasticsearch/http_client.rb:53:in `non_threadsafe_bulk'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.4.1-java/lib/logstash/outputs/elasticsearch/http_client.rb:38:in `bulk'", "org/jruby/ext/thread/Mutex.java:149:in `synchronize'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.4.1-java/lib/logstash/outputs/elasticsearch/http_client.rb:38:in `bulk'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.4.1-java/lib/logstash/outputs/elasticsearch/common.rb:160:in `safe_bulk'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.4.1-java/lib/logstash/outputs/elasticsearch/common.rb:99:in `submit'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.4.1-java/lib/logstash/outputs/elasticsearch/common.rb:84:in `retrying_submit'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.4.1-java/lib/logstash/outputs/elasticsearch/common.rb:28:in `multi_receive'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.0-java/lib/logstash/output_delegator.rb:119:in `worker_multi_receive'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.0-java/lib/logstash/output_delegator.rb:65:in `multi_receive'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.0-java/lib/logstash/pipeline.rb:275:in `output_batch'", "org/jruby/RubyHash.java:1342:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.0-java/lib/logstash/pipeline.rb:275:in `output_batch'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.0-java/lib/logstash/pipeline.rb:206:in `worker_loop'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.0-java/lib/logstash/pipeline.rb:175:in `start_workers'"], :level=>:error}
{:timestamp=>"2016-02-13T20:28:54.532000+0000", :message=>"[400] {\"error\":\"ActionRequestValidationException[Validation Failed: 1: script or doc is missing;2: script or doc is missing;]\",\"status\":400}", :class=>"Elasticsearch::Transport::Transport::Errors::BadRequest", :backtrace=>["/opt/logstash/vendor/bundle/jruby/1.9/gems/elasticsearch-transport-1.0.15/lib/elasticsearch/transport/transport/base.rb:146:in `__raise_transport_error'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/elasticsearch-transport-1.0.15/lib/elasticsearch/transport/transport/base.rb:256:in `perform_request'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/elasticsearch-transport-1.0.15/lib/elasticsearch/transport/transport/http/manticore.rb:54:in `perform_request'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/elasticsearch-transport-1.0.15/lib/elasticsearch/transport/client.rb:125:in `perform_request'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/elasticsearch-api-1.0.15/lib/elasticsearch/api/actions/bulk.rb:87:in `bulk'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.4.1-java/lib/logstash/outputs/elasticsearch/http_client.rb:53:in `non_threadsafe_bulk'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.4.1-java/lib/logstash/outputs/elasticsearch/http_client.rb:38:in `bulk'", "org/jruby/ext/thread/Mutex.java:149:in `synchronize'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.4.1-java/lib/logstash/outputs/elasticsearch/http_client.rb:38:in `bulk'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.4.1-java/lib/logstash/outputs/elasticsearch/common.rb:160:in `safe_bulk'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.4.1-java/lib/logstash/outputs/elasticsearch/common.rb:99:in `submit'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.4.1-java/lib/logstash/outputs/elasticsearch/common.rb:84:in `retrying_submit'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.4.1-java/lib/logstash/outputs/elasticsearch/common.rb:28:in `multi_receive'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.0-java/lib/logstash/output_delegator.rb:119:in `worker_multi_receive'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.0-java/lib/logstash/output_delegator.rb:65:in `multi_receive'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.0-java/lib/logstash/pipeline.rb:275:in `output_batch'", "org/jruby/RubyHash.java:1342:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.0-java/lib/logstash/pipeline.rb:275:in `output_batch'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.0-java/lib/logstash/pipeline.rb:206:in `worker_loop'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.0-java/lib/logstash/pipeline.rb:175:in `start_workers'"], :level=>:warn}
```

After checking the codebase, it seems the ``script`` and ``doc_as_upsert`` configs are mutually exclusive. The infringing code seems to be in [update_action_builder](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch/http_client.rb#L198)

cc @andrewvc 
Thanks

**Comment by kiryam** [comment]
Can you provide your message which you insert into RabbitMQ queue?

**Comment by salimane** [comment]
Here is an example of a message
```
{"id":"1","name":"MY Name","frequency":1}
```
Thanks

**Comment by kiryam** [comment]
If you make update you should add "doc" field eg:
```
{ "update" : {"id" : "1", "_type" : "type1", "_index" : "index1"} }
{ "doc": {"name":"MY Name","frequency":1}}

```

see Elastic bulk API docs https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html

**Comment by salimane** [comment]
@kiryam with the following : 
```
input {
  rabbitmq {
    host => "localhost"
    port => 5672
    vhost => "/"
    user => "my_user"
    password => "my_pwd"
    queue => "my_queue"
    exchange => "my_exchange"
    key => "my_routing_key"
  }
}
output {
  elasticsearch {
    action => "update"
    doc_as_upsert => true
    script_lang => "groovy"
    script_type => "inline"
    document_type => "my_type"
    index => "my_index"
    hosts => ["localhost:9200"]
    document_id => "%{id}"
  }
}
```

everything works as expected. It stops working when I add the ``script`` config.

Thanks

**Comment by kiryam** [comment]
You have this error, because logstash-elasticsearch-output doesn't use strict ElasticSearch bulk api.
I make pull request #357 for fix this. He pending to merge.

Your example:
```
{"id":"1","name":"MY Name","count":1}
```

Works because plugin has lines below:
```
if args[:_script]
    <script_logic>
else
    source = { 'doc' => source }
end
```

This code add "doc" if you doesn't use "script". But if you use script you must do it yourself.

**Comment by salimane** [comment]
@kiryam euh... what do you mean by "But if you use script you must do it yourself." ?
logstash-output-elasticsearch should take care of this, right ? :)
By the way, is there a way to get the event/doc in the elasticsearch output ? like
```
output {
  elasticsearch {
    action => "update"
    upsert => "%{event}" # here
    script_lang => "groovy"
    script_type => "inline"
    document_type => "my_type"
    index => "my_index"
    hosts => ["localhost:9200"]
    document_id => "%{id}"
  }
}
``
Thanks

**Comment by Da-Wei** [comment]
Hi @salimane,
you're right, I thought wrong that "script" and "doc_as_upsert" was mutually exclusive, it's a bug !

**Comment by salimane** [comment]
I've made this PR #367. I think it fixes these issues #357, #364,  #359,  #356, #367

Thanks
